### PR TITLE
change the statusbar commands so that no accidential status bars are created

### PR DESCRIPTION
**Pull request by ailin-nemui**
New command suggestion:

#### Syntax ####

<div class="highlight irssisyntax"><code><span class="synB">STATUSBAR</span> <span class="synB">ADD</span>|<span class="synB">MODIFY</span> <span class="syn10">[<span class="syn">-disable</span> | <span class="syn">-nodisable</span>]</span> <span class="syn10">[<span class="syn">-type</span> <span class="syn">window</span>|<span class="syn">root</span>]</span> <span class="syn10">[<span class="syn">-placement</span> <span class="syn">top</span>|<span class="syn">bottom</span>]</span> <span class="syn10">[<span class="syn">-position</span> <span class="syn">#</span>]</span> <span class="syn10">[<span class="syn">-visible</span> <span class="syn">always</span>|<span class="syn">active</span>|<span class="syn">inactive</span>]</span> <span class="synB05">&lt;statusbar></span></code></div>

<br>

<div class="highlight irssisyntax"><code><span class="synB">STATUSBAR</span> <span class="synB">RESET</span> <span class="synB05">&lt;statusbar></span></code></div>

<br>

<div class="highlight irssisyntax"><code><span class="synB">STATUSBAR</span> <span class="synB">ADDITEM</span>|<span class="synB">MODIFYITEM</span> <span class="syn10">[<span class="syn">-before</span> | <span class="syn">-after</span> <span class="syn09">&lt;item></span>]</span> <span class="syn10">[<span class="syn">-priority</span> <span class="syn">#</span>]</span> <span class="syn10">[<span class="syn">-alignment</span> <span class="syn">left</span>|<span class="syn">right</span>]</span> <span class="synB05">&lt;item></span> <span class="synB05">&lt;statusbar></span></code></div>

<br>

<div class="highlight irssisyntax"><code><span class="synB">STATUSBAR</span> <span class="synB">REMOVEITEM</span> <span class="synB05">&lt;item></span> <span class="synB05">&lt;statusbar></span></code></div>

<br>

<div class="highlight irssisyntax"><code><span class="synB">STATUSBAR</span> <span class="synB">INFO</span> <span class="synB05">&lt;statusbar></span></code></div>

#### Parameters ####

|   |   |
| --- | --- |
| `ADD`: |           Adds a statusbar to the list of statusbars. |
| `MODIFY`: |        Modifies the configuration of a statusbar. |
| `RESET`: |         Restores the default statusbar configuration. |
| `ADDITEM`: |       Adds an item to the specified statusbar. It can be set to appear before/after another item and left/right aligned to a specified position on the screen. |
| `MODIFYITEM`: |    Changes an item position inside a bar. |
| `REMOVEITEM`: |    Removes an item from the specified statusbar. |
| `INFO`: |          List the current details and items of the specified statusbar. |

|   |   |
| --- | --- |
| `-disable`: |      Removes a statusbar from the list. |
| `-type`: |         Sets the type of statusbar, for each split window or only for the current root screen. |
| `-placement`: |    Sets the placement of the statusbar, either at the top or the bottom of the screen. |
| `-position`: |     Sets the position of the statusbar. Represented as a number, with 0 implying the first position. |
| `-visible`: |      Sets the visibility of the statusbar or item. If set to always it is visible on all screens, otherwise if set to inactive or active then it is only visible on inactive or active screens, respectively. |
| `-before`: |       This item is added before the other item. |
| `-after`: |        This item is added after the other item. |
| `-priority`: |     When the statusbar items overflow, the item with the lowest priority is removed first |
| `-alignment`: |    Display the item on the right side. |

Where statusbar refers to the name of the statusbar; if no argument is
given, the entire list of statusbars will be displayed.

### CODE REVIEW

**Comment by ailin-nemui** [comment]
@dequis not sure about `additem` subcommand, thoughts?

**Comment by dequis** [comment]
Before reading too much: hell yeah we needed this, the statusbar commands suck

**Comment by ailin-nemui** [comment]
@irssi/developers if there isn't any further comment I suggest this for merging

File: src/fe-text/statusbar-config.c [hunk 1]
```diff
@@ -463,23 +522,38 @@ static CONFIG_NODE *statusbar_items_section(CONFIG_NODE *parent)
         return parent;
 }
 
-/* SYNTAX: STATUSBAR <name> ADD [-before | -after <item>]
-           [-priority #] [-alignment left|right] <item> */
-static void cmd_statusbar_add(const char *data, void *server,
-			      void *item, CONFIG_NODE *node)
+/* SYNTAX: STATUSBAR ADDITEM|MODIFYITEM [-before | -after <item>]
+           [-priority #] [-alignment left|right] <item> <statusbar> */
```
Comment (dequis) [inline src/fe-text/statusbar-config.c]: Not sure about additem being `<item>` first `<statusbar>` last

File: src/fe-text/statusbar-config.c [hunk 2]
```diff
@@ -334,98 +354,137 @@ static void cmd_statusbar_list(void)
...
+		return;
 	}
-}
 
-/* SYNTAX: STATUSBAR <name> POSITION <num> */
-static void cmd_statusbar_position(const char *data, void *server,
-                                   void *item, CONFIG_NODE *node)
-{
-	iconfig_node_set_int(node, "position", atoi(data));
+	node = sbar_node(name, TRUE);
```
Comment (dequis) [inline src/fe-text/statusbar-config.c]: If i'm reading this right both add and modify call `sbar_node` with create=TRUE. Same thing with additem/modifyitem.

File: src/fe-text/statusbar-config.c [hunk 3]
```diff
@@ -572,17 +697,19 @@ void statusbar_config_init(void)
...
-        command_bind("statusbar type", NULL, (SIGNAL_FUNC) cmd_statusbar_type);
-        command_bind("statusbar placement", NULL, (SIGNAL_FUNC) cmd_statusbar_placement);
-        command_bind("statusbar position", NULL, (SIGNAL_FUNC) cmd_statusbar_position);
-        command_bind("statusbar visible", NULL, (SIGNAL_FUNC) cmd_statusbar_visible);
-
-	command_set_options("statusbar add", "+before +after +priority +alignment");
+        command_bind("statusbar info", NULL, (SIGNAL_FUNC) cmd_statusbar_info);
+        command_bind("statusbar additem", NULL, (SIGNAL_FUNC) cmd_statusbar_additem_modifyitem);
+        command_bind("statusbar modifyitem", NULL, (SIGNAL_FUNC) cmd_statusbar_additem_modifyitem);
+        command_bind("statusbar removeitem", NULL, (SIGNAL_FUNC) cmd_statusbar_removeitem);
```
Comment (dequis) [inline src/fe-text/statusbar-config.c]: space indents

File: src/fe-text/statusbar-config.c [hunk 4]
```diff
@@ -591,13 +718,12 @@ void statusbar_config_deinit(void)
...
-        command_unbind("statusbar add", (SIGNAL_FUNC) cmd_statusbar_add);
-        command_unbind("statusbar remove", (SIGNAL_FUNC) cmd_statusbar_remove);
-        command_unbind("statusbar type", (SIGNAL_FUNC) cmd_statusbar_type);
-        command_unbind("statusbar placement", (SIGNAL_FUNC) cmd_statusbar_placement);
-        command_unbind("statusbar position", (SIGNAL_FUNC) cmd_statusbar_position);
-        command_unbind("statusbar visible", (SIGNAL_FUNC) cmd_statusbar_visible);
+        command_unbind("statusbar info", (SIGNAL_FUNC) cmd_statusbar_info);
+        command_unbind("statusbar additem", (SIGNAL_FUNC) cmd_statusbar_additem_modifyitem);
+        command_unbind("statusbar modifyitem", (SIGNAL_FUNC) cmd_statusbar_additem_modifyitem);
+        command_unbind("statusbar removeitem", (SIGNAL_FUNC) cmd_statusbar_removeitem);
```
Comment (dequis) [inline src/fe-text/statusbar-config.c]: space indents

File: src/fe-text/statusbar-config.c [hunk 5]
```diff
@@ -463,23 +522,38 @@ static CONFIG_NODE *statusbar_items_section(CONFIG_NODE *parent)
...
+		printformat(NULL, NULL, MSGLEVEL_CLIENTERROR,
+			    TXT_STATUSBAR_NOT_FOUND, statusbar);
+		cmd_params_free(free_arg);
 		return;
+	}
+
+        node = statusbar_items_section(node);
+	if (node == NULL) {
+		cmd_params_free(free_arg);
+                return;
```
Comment (dequis) [inline src/fe-text/statusbar-config.c]: space indents

File: src/fe-text/statusbar-config.c [hunk 6]
```diff
@@ -334,98 +354,137 @@ static void cmd_statusbar_list(void)
...
+			    TXT_STATUSBAR_NOT_FOUND, name);
+		cmd_params_free(free_arg);
+		return;
 	}
+
+	parent = iconfig_node_traverse("statusbar", TRUE);
+	parent = iconfig_node_section(parent, active_statusbar_group->name,
+				     NODE_TYPE_BLOCK);
+
+        iconfig_node_set_str(parent, node->key, NULL);
```
Comment (dequis) [inline src/fe-text/statusbar-config.c]: space indents

File: src/fe-text/statusbar-config.c [hunk 7]
```diff
@@ -273,6 +274,25 @@ static const char *sbar_get_visibility(STATUSBAR_CONFIG_REC *rec)
...
+{
+        CONFIG_NODE *node;
+	STATUSBAR_CONFIG_REC *rec = statusbar_config_find(active_statusbar_group, name);
+	if (rec != NULL) {
+		name = rec->name;
+	} else if (!create) {
+		return NULL;
+	}
+
+        /* lookup/create the statusbar node */
```
Comment (dequis) [inline src/fe-text/statusbar-config.c]: space indents

File: src/fe-text/statusbar-config.c (hunk 1 above)
Comment (ailin-nemui) [reply src/fe-text/statusbar-config.c]: it somewhat matches `/channel add <channel> <network>`

File: src/fe-text/statusbar-config.c (hunk 7 above)
Comment (ailin-nemui) [reply src/fe-text/statusbar-config.c]: .. I thought I had ran the almight git-clang-format  ..

File: src/fe-text/statusbar-config.c [hunk 8]
```diff
@@ -502,67 +576,118 @@ static void cmd_statusbar_add(const char *data, void *server,
...
+	if (node == NULL) {
+		printformat(NULL, NULL, MSGLEVEL_CLIENTERROR,
+			    TXT_STATUSBAR_NOT_FOUND, statusbar);
+		cmd_params_free(free_arg);
+		return;
+	}
+
         node = statusbar_items_section(node);
 	if (node == NULL)
                 return;
```
[the remark this answers is not in the document]
Comment (ailin-nemui) [reply src/fe-text/statusbar-config.c]: good catch

File: src/fe-text/statusbar-config.c (hunk 2 above)
Comment (ailin-nemui) [reply src/fe-text/statusbar-config.c]: oops

**Comment by ailin-nemui** [comment]
any suggestion as to where/which position to mention the change in syntax?

File: src/fe-text/statusbar-config.c (hunk 1 above)
Comment (dequis) [reply src/fe-text/statusbar-config.c]: I guess. No strong opinion either way.

**Comment by dequis** [comment]
I was thinking a single-sentence mention in the Description section of the /help, without going in detail

**Comment by ailin-nemui** [comment]
@dequis wow that was more complicated than I thought. it could need another look. Basically the old methods were hard-wired to always create and copy the status bar items, so I had to change that with tests for the internal/default config and only copying if we actually are about to change something

**Comment by ailin-nemui** [comment]
@irssi/developers

**Pull request merged**